src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp

```diff
@@ -0,0 +1,228 @@
+
+    {
+        const auto b_const = std::dynamic_pointer_cast<v0::Constant>(b_quantized.get_node_shared_ptr());
+
+        ov::Output<ov::Node> casted_b;
+        ov::Shape casted_b_shape;
+        ov::Output<ov::Node> default_zp;
+        // Casting/converting data of source constant.
+        // For further calculations (sub and/or multiply) we need to reshape it from [N][n_blocks_per_col][blob_size *
+        // X] to [N * n_blocks_per_col][blob_size * X] (where X is amount of values in 1 byte) because scale and
+        // zero_point are represented as: ...with shape like: [N * n_blocks_per_col]...
+        switch (bits) {
+        case 2:
+            casted_b_shape = ov::Shape{static_cast<size_t>(N * n_blocks_per_col), static_cast<size_t>(blob_size * 4)};
+            casted_b = std::make_shared<v0::Constant>(ov::element::u2, casted_b_shape, b_const->get_data_ptr());
+            default_zp = std::make_shared<v0::Constant>(a.get_element_type(), Shape{}, 2);
+            break;
+        case 4:
+            casted_b_shape = ov::Shape{static_cast<size_t>(N * n_blocks_per_col), static_cast<size_t>(blob_size * 2)};
+            casted_b = std::make_shared<v0::Constant>(ov::element::u4, casted_b_shape, b_const->get_data_ptr());
+            default_zp = std::make_shared<v0::Constant>(a.get_element_type(), Shape{}, 8);
+            break;
+        case 8:
+            casted_b_shape = ov::Shape{static_cast<size_t>(N * n_blocks_per_col), static_cast<size_t>(blob_size)};
+            casted_b = op::util::reshape(b_const, casted_b_shape);
+            default_zp = std::make_shared<v0::Constant>(a.get_element_type(), Shape{}, 128);
+            break;
+        default:
+            FRONT_END_THROW("Unsupported bits count");
+            break;
+        }
+
+        // Possible issue with slice implementation, had to move convertion before slice, instead of slicing uint4
+        const auto converted_b = std::make_shared<v1::ConvertLike>(casted_b, a);
+
+        // Simple case
+        if (n_blocks_per_col == 1) {
+            // Removing unused items in case block is bigger than column count
+            const auto zero_const = std::make_shared<v0::Constant>(ov::element::i32, Shape{1}, 0);
+            const auto one_const = std::make_shared<v0::Constant>(ov::element::i32, Shape{1}, 1);
+            const auto elements_const =
+                std::make_shared<v0::Constant>(ov::element::i32, Shape{1}, static_cast<int32_t>(K));
+            const auto axis_const = std::make_shared<v0::Constant>(ov::element::i32, Shape{1}, 1);
+            const auto slice_b =
+                std::make_shared<v8::Slice>(converted_b, zero_const, elements_const, one_const, axis_const);
+
+            // Transpose matrix
+            const auto transposed_shape =
+                std::make_shared<v0::Constant>(ov::element::i64, Shape{2}, std::vector<int64_t>{1, 0});
+            const auto transposed_b = std::make_shared<v1::Transpose>(slice_b, transposed_shape);
```

Comment:
This transpose runs in low bit precision: `u2`/`u4`. It may have the same issues as `Slice`. Have you checked the correctness of this particular step?

Then I'm not quite sure what this `Transpose` gives in terms of the original matrix B. It looks to permute also the blocked dimension, but due to that multiplier that we applied when creating `converted_b` depending on the type of elements. Could you provide some visuals or extended description, or link where it is described in more detailed form in comparison to the original ORT spec?